graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/SocketModuleBuiltins.java

```diff
@@ -113,4 +210,79 @@ String doGeneric() {
             }
         }
     }
+
+    @Builtin(name = "getaddrinfo", parameterNames = {"host", "port", "family", "type", "proto", "flags"})
+    @GenerateNodeFactory
+    public abstract static class GetAddrInfoNode extends PythonBuiltinNode {
+        @Specialization
+        @TruffleBoundary
+        Object getAddrInfo(String host, int port, int family, int type, int proto, PInt flags) {
+            return getAddrInfo(host, port, family, type, proto, flags.asInt());
+        }
+        @Specialization
+        @TruffleBoundary
+        Object getAddrInfo(String host, String port, int family, int type, int proto, PInt flags) {
+            return getAddrInfo(host, port, family, type, proto, flags.asInt());
+        }
+        @Specialization
+        @TruffleBoundary
+        Object getAddrInfo(String host, int port, int family, int type, int proto, int flags) {
+            InetAddress[] adresses = resolveHost(host);
+            List<Service> serviceList = new ArrayList();
+            serviceList.add(new Service(port, "tcp"));
+            serviceList.add(new Service(port, "udp"));
+            return mergeAdressesAndServices(adresses, serviceList, proto, flags);
+        }
+        @Specialization
+        @TruffleBoundary
+        Object getAddrInfo(String host, String port, int family, int type, int proto, int flags) {
+            if (services == null) {
+                services = parseServices();
+            }
+            List<Service> serviceList = services.get(port);
+
+            InetAddress[] adresses = resolveHost(host);
+            return mergeAdressesAndServices(adresses, serviceList, proto, flags);
+        }
+
+        private Object mergeAdressesAndServices(InetAddress[] adresses, List<Service> serviceList, int proto, int flags) {
```

Comment:
This needs a TruffleBoundary (uses ArrayList). My general rule is to move the TruffleBoundary to the most specific method possible, so all the callers don't need it